[Amy Lim Zhi Ting] iP

amylzting opened the pull request.

Commit:
Add Gradle support

amylzting committed:
Add skeletal version of Dino

Let's add a greeting that Dino says each time he starts,
and configure Dino such that he echos the user command
and outputs a horizontal line separating each output,
and exits when the user types bye.

Using a horizontal line to separate each output from Dino
improves the visibility.

amylzting committed:
Add the storage ability to Dino

Let's add the ability for Dino to store whatever text
entered by the user and display them back to
the user when requested, in an indexed list format.

amylzting committed:
Add ability for Dino to mark tasks as done

Let's configure Dino such that he marks tasks
as done when user inputs in format
"done <task number>", and displays whether
tasks are done in task list.
Let's implement a Task class to represent
tasks.

amylzting committed:
Add support for three types of Task in Dino

Let's implement 3 subclasses of Task, namely:
Todo, Deadline and Event. Todos are tasks without
any date/time, Deadlines are tasks that need to be
done before a specific date/time, and Events are
tasks that start at a specific time and ends at a specific time.

Inheritance is used to support multiple task types,
as they are all tasks and there is similarity between them.

amylzting committed:
Test using the I/O redirection technique

Let's,
* use the input/output redirection technique
  to semi-automate the testing of Duke. Check
  all cases by using all Dino commands in
  input.txt.
* update Dino's introduction, fix
  whitespaces issues and fix bug where scanner
  cannot find next line.
* fix issue where symbol ✓ and ✘ outputs
  as ? when running runtest.bat.

amylzting committed:
Refactoring Duke.java code

Let's refactor the main method into
multiple smaller methods.
Let's add DukeExceptions class.

Refactoring makes the code neater
and easier to maintain.

amylzting committed:
Add DukeException error handling to Dino

Let's handle errors that occur for each feature
in Dino by throwing DukeException.
Let's update the test cases in input.txt,
to better test the functionality.

DukeException represent exceptions specific to Dino.

amylzting committed:
Add ability for Dino to delete tasks

Let's add support for deleting tasks from the list.
Let's add more testcases in input.txt to test the
functionality of the deleteTask function.

amylzting committed:
Update Dino.java user instruction

Let's add clearer instructions for the
commands that can be used in Dino.

Clearer instructions help users to use
Dino more easily.

amylzting committed:
Update Dino user instructions

Let's add an instruction teaching users
how to exit from Dino in the introduction.

This teaches users how to exit from the
program and improves user experience.

amylzting committed:
Initialise DukeAction enum

Let's initialise a DukeAction enum,
to represent the actions that Dino takes,
such as marking task as done,
adding task to list and deleting task from
list.

Creating an enum makes it easier to
reference the action when updating
data in task list.

amylzting committed:
Add ability for Dino to load task data

Let's add the ability for Dino to:
* create a directory data to store user data
* create a file to store user task list
* load data from user task list

This helps to keep track of user's task data
when they use Dino again.

amylzting committed:
Add ability to update changes into hard disk

Let's add the ability for Dino to update the
tasklist.txt file whenever a task is added,
deleted, or marked as done.
Let's change the tick and cross symbol
denoting task done to 1 and 0.

amylzting committed:
Change markAsDone symbol in Task class

Let's change the symbol denoting whether a
task is done from tick and cross to 1 and 0.

This takes away any potential bugs caused
from symbol.

amylzting committed:
Change markAsDone symbol in Task class

Let's change the symbol denoting whether a
task is done from tick and cross to 1 and 0.

This takes away any potential bugs caused
from symbol.

amylzting committed:
Add ability for Dino to recognise dates and times

Let's
* only accept yyyy-mm-dd as date format and
hhmm as time format for user input, for deadline
and event tasks.
* implement method in Deadline and Event classes
to format date and time of deadline/event into
recognisable syntaxes.

amylzting committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Task.java

Let's resolve the comment conflict.

amylzting committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Task.java

Let's resolve conflicts.

amylzting committed:
Fix bugs occured from merging

Let's fix the bugs that occurred when
merging branch-Level-7 and branch-Level-8.

As formats used for the deadline times for
the tasks are different (due to addition of
Level 8 functionality), updating the methods
used in Level 7 and Level 8, will solve any
issues that disrupted the functionality.

amylzting committed:
Change the format of tasks stored in hard disk

Need to change the format of the tasks stored in
hard disk for it to be easier to process when
Dino starts.

Let's change the format of tasks stored in hard
disk. Task information are separated by and '@'
separator.

amylzting committed:
Add clases Ui, Storage, Parser, TaskList

More classes need to be added to make
code more OOP.

Let's create classes Ui, Storage, Parser and
TaskList to refactor the code to extract out
closely related code as classes and
make code more OOP.

amylzting committed:
Implement Storage, TaskList, Ui, Parser classes

Let's implements the Storage, TaskList, Ui and
Parser classes to refactor code from Duke.java
into different classes.

amylzting committed:
Add abstract class Command and its subclasses

Let's
* make our code more OOP and implement
an abstract class Command which handles user
commands.
* implement subclasses of Command to represent
each different command from the user.
* update the Ui class
* update the Parser class to parse commands

Using a Command class to represent user commands
improves the OOP of the entire code.

amylzting committed:
Modify Ui class

Ui should show errors that occur.

Let's add a method in Ui class to show error
messages.

Showing error messages tells users what
went wrong with the program.

amylzting committed:
Update filepath variable after code refactor

Let's update the hardcoded file path in
writeToFile method in Storage class, to
the filepath variable.

amylzting committed:
Divide classes into packages

Let's organize the classes into suitable java packages.

Organizing the classes into packages give easier
management.

amylzting committed:
Add JUnit tests

Need to add tests to test
the functionality of the methods.

Let's add JUnit test to test
* the behaviour of Deadline and Event creation.
* the behaviour of converting tasks to
stored string format to store
* the exception thrown during invalid Event creation.

amylzting committed:
Update JUnit tests and fix bug

Need to add more comprehensive JUnit
test for the creation of Deadline and Event
tasks on wrong user input.

Let's
* add more test cases in testing the exception
thrown in case of wrong user input in creation
of tasks.
* fix ArrayOutOfBounds exception thrown in
task creation in Deadline and Event.

amylzting committed:
Add JavaDoc comments

There is currently no JavaDoc comments
on the methods of all the classes.

Let's add Javadoc comments to all the
methods and constructors that are not
self-explanatory.

Adding javadoc comments makes the code
clearer and helps others to understand the code.

amylzting committed:
Update code to follow Java Coding Standard

Some parts of the code are not adhering
to the Java coding standard.

Let's update some of the code indentation,
white spaces and import statements to
adhere to the java coding standard.

Adhering to the java coding standards makes
the code neater and more uniform.

amylzting committed:
Add ability for Dino to find tasks from a keyword

Dino does not have the ability to search for tasks.

Let's add an ability for Dino to search for tasks based
on a keyword entered by user and list out
all the matching tasks.

amylzting committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/tasks/TaskList.java
#	text-ui-test/data/tasklist.txt

amylzting committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/tasks/TaskList.java
#	text-ui-test/data/tasklist.txt

amylzting committed:
Resolve merge conflict in Command class

amylzting committed:
Updated runtest.bat file

The file did not have command to delete
tasklist.txt each time it is ran and hence,
I had to manually delete tasklist.txt each time
I wanted to test. Also, it did not sync properly
with my entire code due to packages.

Let's add and change up some commands in
the runtest.bat file, and add more testcases to
test the functionality of the code.

Doing this, tasklist.txt is automatically deleted
each time i run the test.

amylzting committed:
Update javadocs for methods associated with find

Javadocs have not been written for execute method
in FindCommand class and findTask method
in TaskList class.

Let's add javadocs for the relevant methods added
for the find functionality in Dino.

amylzting committed:
Merge remote-tracking branch 'origin/add-gradle-support'

amylzting committed:
Update and add builds in build.gradle

mainClassName and checkstyle version
needs to be updated.

Let's update mainClassName and
checkstyle version, and enable assertions.

amylzting committed:
Set up checkstyle plugin

The project does not have checkstyle
plugin installed.

Let's
* install checkstyle plugin to
automatically check the style of the
code when writing.
* correct some of the code style.

amylzting committed:
Merge branch 'branch-A-Gradle'

amylzting committed:
Merge branch 'branch-A-CheckStyle'

amylzting committed:
Add javafx to build.gradle

Let's update build.gradle so we can
add javafx to the project.

saad-haider-pp commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,67 @@
+
+
+
+public class Duke {
```
I like this! Follows the coding style guide almost to the tee 😄

saad-haider-pp commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,67 @@
+    public static void main(String[] args) {
+        new Duke("./data/tasklist.txt").run();
+    }
+}
```
Not to be pedantic, but to keep in line with the coding standard, should there be a new line at the end? I've seen similar things in other files as well.

amylzting replied:
Done!

saad-haider-pp commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,25 @@
+import duke.tasks.TaskList;
+import duke.ui.Ui;
+
+public class ByeCommand extends Command {
+
+    public ByeCommand() {
+        this.isBye = true;
+    }
```
Any reason why the class and constructor do not have javadocs? I noticed the same thing in other files too.

amylzting replied:
Have added the javadocs in recent commit!

saad-haider-pp commented on src/main/java/duke/tasks/Todo.java:
```diff
@@ -0,0 +1,17 @@
+package duke.tasks;
+
+public class Todo extends Task {
+
+    public Todo(String description) {
+        super(description);
+    }
+
+    @Override
+    public String toString() {
+        return "[T]" + super.toString();
+    }
+
+    public String storedTaskString() {
+        return "T" + "@" + super.storedTaskString();
+    }
+}
```
Not sure I like the absence of javadocs here

amylzting replied:
Have added the javadocs in recent commit!

saad-haider-pp reviewed (commented):
Overall, it's almost LGTM. It's evident that a clear and deliberate effort was made to adhere to the coding standard, and with just a few nits to fix, it will be set.

amylzting committed:
Set up basic GUI features

Let's set up GUI for Dino using
fxml by creating DialogBox.fxml and
MainWindow.fxml, and classes to
launch the application.

amylzting committed:
Merge branch 'branch-Level-10'

okg00dbye commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    /**
+     * Initialises Duke with the file record of task history, and runs Duke.
```
empty line between description and parameter sections

amylzting replied:
Done!

okg00dbye commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,202 @@
+     * @return List list of Task stored in hard disk.
+     * @throws DukeException If could not load task information from hard disk.
+     */
+    public List<Task> loadData() throws DukeException {
```
method is a bit long, maybe can separate the first file creation part into another method?

okg00dbye commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,202 @@
+                    if (tempFile.renameTo(inputFile)) {
+                        System.out.println("Success!");
+                    } else {
+                        throw new DukeException("Dino could not write task data to hard disk.");
```
nice error catching :-)

amylzting replied:
thank you :)

okg00dbye commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,100 @@
+import duke.DukeException;
+
+
+public class Deadline extends Task {
```
header comment for public class (for event/task/tasklist ..)

amylzting replied:
done in recent commit!

okg00dbye commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,77 @@
+                + "|____/|_|_| |_|\\___/\n";
+        System.out.println("Rawr! I'm Dino ><\n"
+                + logo
+                + "\nGet started on your task list by entering a task!"
```
i like the tutorial!! nice hehe

amylzting replied:
thank you:)

okg00dbye commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Test
+    public void createDeadline_wrongDateTimeFormat_exceptionThrown() {
```
nice test names :-)

amylzting replied:
thank you:)

okg00dbye commented on src/main/java/duke/command/FormatCommand.java:
```diff
@@ -0,0 +1,47 @@
+     * @return String of input format information
+     */
+    @Override
+    public String executeToString(Storage storage, TaskList taskList, Ui ui) {
```
ah nice, very thoughtful for user hehe

amylzting replied:
thanks!

okg00dbye commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,202 @@
+     * Initialises a storage object that stores any updates to the
+     * list of tasks in the hard disk and retrieves task information
+     * from the hard disk when Duke starts.
+     * @param filepath string file path of the file where Duke stores
```
empty line

amylzting replied:
Done in recent commit!

okg00dbye commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,202 @@
+                break;
+
+            case MARK_DONE:
+                File taskFile = new File(this.filepath);
```
maybe can abstract file creation and reader/writer creation to before the switch statement?

amylzting replied:
thanks for the suggestion, done in recent commit!

okg00dbye commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,211 @@
+                    // task deadline taken as string after first '/at'
+                    Event event = Event.createEvent(eventAt[0], eventAt[1].substring(1));
+                    this.taskList.add(event);
+                    storage.writeToFile(event, DukeAction.ADD);
```
maybe storage.writeToFile and success statement can be abstracted to outside of the switch statement also? can improve readability :-)

okg00dbye reviewed (commented):
very clean code and well documented!! :-) a few methods are a bit long and can be abstracted a bit to improve readability but its not a big thing. i especially like the additional touches like guiding the user how to use dino! (also what are the odds that i was assigned you hoho hi Amy :D )

amylzting committed:
Update some javadocs and coding style

amylzting committed:
Update some Ui texts

amylzting committed:
Update format of task string stored

Let's update the format of the task string
to be stored in the task list in hard disk
and update timeFormat function to catch
NumberFormatException.

amylzting committed:
Add assertions

Some parts of code require some assertions.

Let's add assertions some of the private methods
such as timeFormat.

amylzting committed:
Update code quality

Some of the code quality needs to
be improved.

Let's
* refactor the code.
* follow the SLAP principle.

amylzting committed:
Merge pull request #1 from amylzting/branch-A-Assertions

Add assertions

amylzting committed:
Merge branch 'master' of https://github.com/amylzting/ip

* 'master' of https://github.com/amylzting/ip:
  Add assertions

amylzting committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/tasks/TaskList.java

amylzting committed:
Merge pull request #2 from amylzting/branch-A-CodeQuality

Update code quality

amylzting committed:
Merge branch 'master' of https://github.com/amylzting/ip

* 'master' of https://github.com/amylzting/ip:
  Update code quality

amylzting committed:
Add Priority to tasks

Let's
* create Priority enum.
* add Priority to all tasks.

amylzting committed:
Add set priority functionality to tasks

Let's modify the code such that user
is able to set the priority of tasks and the
change in priority will be reflected in hard disk.

amylzting committed:
Update Ui to show setting of priority

Let's update the format ui to teach users
how to set priority of tasks.

amylzting committed:
Improve the GUI of Dino

Let's distinguish between user dialog
and Dino's dialog and make overall
GUI appearance more appealing.

amylzting committed:
Change format command to help command

Let's change the name of 'format' command of Dino
to 'help' command of Dino for increased readability and
understanding of the command.

amylzting committed:
Set theme jekyll-theme-hacker

amylzting committed:
Add user guide for Dino

Let's update README.md in docs
to complete the user guide for Dino.

amylzting committed:
Merge branch 'master' of https://github.com/amylzting/ip

amylzting committed:
Set theme jekyll-theme-midnight

amylzting committed:
Set theme jekyll-theme-dinky

amylzting committed:
Set theme jekyll-theme-tactile

amylzting committed:
Update Dino user guide

amylzting committed:
Set theme jekyll-theme-hacker

amylzting committed:
Set theme jekyll-theme-slate

amylzting committed:
Update Dino user guide

amylzting committed:
Merge branch 'master' of https://github.com/amylzting/ip

amylzting committed:
Rename Ui image to Ui.png

amylzting committed:
Update Ui and screenshot of Dino

amylzting committed:
Reorganize and shorten code for storage in Dino

Let's shorten the code for updating the storage
of task information when a command is
executed.
